### PR TITLE
fix data_norm op to avoid impractical normalization result test=develop

### PR DESCRIPTION
**Pull request by yaoxuefeng6**
In data_norm op, which is usually used with pslib mode. we force to change the distribution of embedding by the historical mean and square sum without doing scale and shift. So in some case, when the slot is empty or new, the value after normalization may be impractical.
To fix this bug, we determine to do this normalization  by judging if the show num of slots is zero.  if  zero, we skip the forward normalization and backward update of these statistic values

### CODE REVIEW

**Review by fuyinno4 — DISMISSED** [review]

**Review by frankwhzhang — DISMISSED** [review]

**Review by fuyinno4 — DISMISSED** [review]

File: paddle/fluid/operators/data_norm_op.cc [hunk 1]
```diff
@@ -125,6 +125,7 @@ class DataNormOpMaker : public framework::OpProtoAndCheckerMaker {
           PADDLE_ENFORCE(epsilon >= 0.0f && epsilon <= 0.001f,
                          "'epsilon' should be between 0.0 and 0.001.");
         });
+    AddAttr<int>("slot_dim", "").SetDefault(-1);
```
Comment (Superjomn) [inline paddle/fluid/operators/data_norm_op.cc]: 加注释，这个是分布式特有的还是单机也需要，有什么用处

File: paddle/fluid/operators/data_norm_op.cc [hunk 2]
```diff
@@ -349,24 +379,58 @@ class DataNormGradKernel<platform::CPUDeviceContext, T>
...
-            batch_square_sum->data<T>(), C);
-        Eigen::Array<T, Eigen::Dynamic, 1> sample_sum(C);
-        Eigen::Array<T, Eigen::Dynamic, 1> sample_square_sum(C);
-        // calculate data sample sum and square sum
-        sample_sum.setZero();
-        sample_square_sum.setZero();
-        for (int nc = 0; nc < N; ++nc) {
-          sample_sum += x_arr.col(nc);
-          sample_square_sum += (x_arr.col(nc) - means_arr).square();
+        if (slot_dim > 0 && N > 0) {
```
Comment (Superjomn) [inline paddle/fluid/operators/data_norm_op.cc:388]: 加一些注释讲明在干啥，下面接了一大段代码

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -2708,6 +2709,8 @@ def data_norm(input,
         moving_variance_name(string, Default None): The name of the moving_variance which store the global Variance.
         do_model_average_for_mean_and_var(bool, Default True): Whether parameter mean and variance
             should do model average when model average is enabled.
+        slot_dim(int): The dimention of one slots used for judging if the show num is zero to skip
```
Comment (Superjomn) [inline python/paddle/fluid/layers/nn.py]: one slots?
if the show num is zero to skip normalization

不通顺， num -> number,
show num 描述更清晰一些，看不懂的话，之前文档白修了

File: python/paddle/fluid/tests/unittests/test_data_norm_op.py [hunk 4]
```diff
@@ -24,14 +24,23 @@
...
+    if slot_dim <= 0:
+        for i in range(x_shape[0]):
+            x[i] -= means_arr
+            x[i] *= scales_arr
+        y = np.array(x)
+    else:
+        y = np.zeros(x_shape).astype(np.float32)
+        for i in range(x_shape[0]):
+            for j in range(0, x_shape[1], slot_dim):
+                if x[i][j] <= -1e-7 or x[i][j] > 1e-7:
```
Comment (Superjomn) [inline python/paddle/fluid/tests/unittests/test_data_norm_op.py]: 1e-7 用 const variable 表示

File: python/paddle/fluid/tests/unittests/test_data_norm_op.py [hunk 5]
```diff
@@ -69,6 +78,7 @@ def check_with_place(self, place, data_layout, dtype, shape):
             data_layout(str): NCHW or NWHC
             dtype(dtype): np.float32
             shape(list): input shape
+            slot_dim(int): dim for one slot
```
Comment (Superjomn) [inline python/paddle/fluid/tests/unittests/test_data_norm_op.py]: slot 是啥意思？ 
dim -> dimension，文档里需要是正常英文表示

**Review by Superjomn — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/data_norm_op.cc (hunk 1 above)
Comment (yaoxuefeng6) [reply paddle/fluid/operators/data_norm_op.cc]: added

File: paddle/fluid/operators/data_norm_op.cc (hunk 2 above)
Comment (yaoxuefeng6) [reply paddle/fluid/operators/data_norm_op.cc:388]: added

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (yaoxuefeng6) [reply python/paddle/fluid/layers/nn.py]: fixed

File: python/paddle/fluid/tests/unittests/test_data_norm_op.py (hunk 4 above)
Comment (yaoxuefeng6) [reply python/paddle/fluid/tests/unittests/test_data_norm_op.py]: fixed

File: python/paddle/fluid/tests/unittests/test_data_norm_op.py (hunk 5 above)
Comment (yaoxuefeng6) [reply python/paddle/fluid/tests/unittests/test_data_norm_op.py]: fixed

**Review by Superjomn — APPROVED** [review]
LGTM

**Review by frankwhzhang — APPROVED** [review]

**Review by fuyinno4 — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]
LGTM

**Pull request merged**